### PR TITLE
More works on backend part

### PR DESCRIPTION
**Pull request by edisonhello**
A `BufferedFirestoreContainer` extends `FirestoreContainer` that will merge data into a big data.
A `SafetyStorage` that makes sure all VSCode share the same memento contents.

### CODE REVIEW

**Comment by edisonhello** [comment]
#293 may be clone if this has no problem.

File: dbux-projects/src/backend/Backlog.js [hunk 1]
```diff
@@ -1,73 +1,112 @@
 import { newLogger } from '@dbux/common/src/log/logger';
+import { isEqual } from 'lodash';
```
Comment (MiccWan) [inline dbux-projects/src/backend/Backlog.js:2]: change to `import isEqual from 'lodash/isEqual';`?

File: dbux-projects/src/backend/Backlog.js [hunk 2]
```diff
@@ -1,73 +1,112 @@
...
-        this.backlog.shift();
-        await this.save();
+        backlog.shift();
       }
       catch (err) {
-        this.backlog.shift();
-        await this.save();
+        backlog.shift();
         warn(`Removed write request after error: ${err.message}\n\n${JSON.stringify(writeRequest)}`);
       }
```
Comment (MiccWan) [inline dbux-projects/src/backend/Backlog.js:84]: The storage will be locked before the loop, no need to worry about new data been pushed while writing, thus we can make this loop O(1) (use for-of or forEach) instead.

File: dbux-projects/src/backend/BufferedFirestoreContainer.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    this.buffer = new SafetyStorage(this._keyName);
+    this._previousFlushTime = new Date();
+
+    Verbose && debug(`restore buffer @${this.collectionName}`, this.buffer.get());
+  }
+
+  /**
+   * @return {Array}
+   */
+  saveGetBuffer() {
```
Comment (MiccWan) [inline dbux-projects/src/backend/BufferedFirestoreContainer.js:28]: safe?

File: dbux-projects/src/backend/db.js [hunk 4]
```diff
@@ -196,10 +197,13 @@ export class Db {
 
       const { collection } = container;
       doc = collection.doc(id);
-      debug('data', data);
+      Verbose && debug('writting data', data);
```
Comment (MiccWan) [inline dbux-projects/src/backend/db.js:200]: typo: writing

**Review by MiccWan — CHANGES_REQUESTED** [review]

**Comment by Domiii** [comment]
needs `try/finally`

**Comment by Domiii** [comment]
backlog needs to identify entries by their unique key, that is a combo key of: `collectionName` and `docId`.

However, neither id nor collection name are stored inside the object to be stored; so you have to find a way to store that data in `FirestoreContaienr` when writing to the `Db` object.

**Comment by edisonhello** [comment]
Also cover #307

**Review by MiccWan — APPROVED** [review]

**Pull request merged**